Source/Core/DolphinQt/Debugger/CodeDiffDialog.cpp

```diff
@@ -0,0 +1,411 @@
+  layout->addWidget(m_reset_btn);
+
+  setLayout(layout);
+}
+
+void CodeDiffDialog::ConnectWidgets()
+{
+  connect(m_record_btn, &QPushButton::toggled, this, &CodeDiffDialog::OnRecord);
+  connect(m_include_btn, &QPushButton::pressed, [this]() { Update(true); });
+  connect(m_exclude_btn, &QPushButton::pressed, [this]() { Update(false); });
+  connect(m_matching_results_list, &QListWidget::itemClicked, [this](QListWidgetItem* item) {
+    m_code_widget->SetAddress(item->data(Qt::UserRole).toUInt(),
+                              CodeViewWidget::SetAddressUpdate::WithUpdate);
+  });
+  connect(m_reset_btn, &QPushButton::pressed, this, &CodeDiffDialog::ClearData);
+  connect(m_matching_results_list, &CodeDiffDialog::customContextMenuRequested, this,
+          &CodeDiffDialog::OnContextMenu);
+}
+
+void CodeDiffDialog::ClearData()
+{
+  if (m_record_btn->isChecked())
+    m_record_btn->toggle();
+  ClearBlockCache();
+  m_matching_results_list->clear();
+  m_exclude_size_label->setText(QStringLiteral("Excluded: 0"));
+  m_include_size_label->setText(QStringLiteral("Included: 0"));
+  m_exclude_btn->setEnabled(false);
+  m_include_btn->setEnabled(false);
+  std::vector<Diff>().swap(m_include);
+  std::vector<Diff>().swap(m_exclude);
+  JitInterface::SetProfilingState(JitInterface::ProfilingState::Disabled);
+}
+
+void CodeDiffDialog::ClearBlockCache()
+{
+  Core::State old_state = Core::GetState();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Paused);
+
+  JitInterface::ClearCache();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Running);
+}
+
+void CodeDiffDialog::OnRecord(bool enabled)
+{
+  JitInterface::ProfilingState state;
+
+  if (enabled)
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Recording..."));
+    state = JitInterface::ProfilingState::Enabled;
+    m_exclude_btn->setEnabled(true);
+    m_include_btn->setEnabled(true);
+  }
+  else
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Start Recording"));
+    state = JitInterface::ProfilingState::Disabled;
+    m_exclude_btn->setEnabled(false);
+    m_include_btn->setEnabled(false);
+  }
+
+  m_record_btn->update();
+  JitInterface::SetProfilingState(state);
+}
+
+void CodeDiffDialog::OnInclude()
+{
+  std::vector<Diff> current_diff;
+  auto current_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_include = current_symbols;
+    return;
+  }
+
+  if (m_exclude.empty())
+  {
+    current_diff = current_symbols;
+  }
+  else
+  {
+    for (auto& iter : current_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+      }
+    }
+  }
+
+  if (!m_include.empty())
+  {
+    // Compare include with current and remove items that aren't in both.
+    std::vector<Diff> tmp_swap;

```

Comment:
Is there a reason not to do this operation on `m_include` directly? Can't the erase-remove idiom or similar be used here?

Reply:
Original:
```
if (!m_include.empty())
  {
    // Compare include with current and remove items that aren't in both.
    std::vector<Diff> tmp_swap;
    for (auto& iter : m_include)
    {
      if (std::any_of(current_diff.begin(), current_diff.end(),
                      [&](Diff& v) { return v.symbol == iter.symbol; }))
      {
        tmp_swap.push_back(iter);
      }
    }
    m_include.swap(tmp_swap);
  }
```

If we act on m_include directly I should inverse the lambda conditional to: 
`return v.symbol != iter.symbol`

However I think I realized why TryTwo uses a tmp object:
If you perform an erase/remove on m_include within the for an index is skipped
ex:
```
m_include = [ 0, 1, 2 , 3, 4 ]
for m_include {
... //match case [match = 2]
      m_include.erase(match)
}
```
->
next loop will be 4 as all succeeding indices shift left.

Reply:
I was rather thinking of the erase-remove idiom (cf. [examples](https://en.cppreference.com/w/cpp/algorithm/remove))
```c++
m_include.erase(
  std::remove_if(m_include.begin(), m_include.end(), RemovePredicate),
  m_include.end()
)
```

Unless I'm mistaken the predicate you would need would be something like that:
```c++
// The argument v is each Diff from m_include
// Remove value from m_include if not in current_diff
[&](Diff& v) { return std::find(current_diff.begin(), current_diff.end(), v) == current_diff.end(); }))
```

Reply:
`[&](Diff& v) { return std::find(current_diff.begin(), current_diff.end(), v) == current_diff.end(); }))`
As for this predicate, I'm not following why `== current_diff.end()`. Is here at all?

Untested, but this is what I went with, using a `find_if` to return true in comparing:

```c++
m_include.erase(std::remove_if(m_include.begin(), m_include.end(),
                                   [&](Diff& v) {
                                     return std::find_if(
                                         current_diff.begin(), current_diff.end(),
                                         [&](Diff& p) { return p.symbol == v.symbol; });
                                   }),
                    m_include.end());```

Reply:
Violates template for  find_if.
Side note, I think if this was valid it would actually remove matches, which is not what we want.
We only want to remove non-matches

Reply:
> `== current_diff.end()`

Means it arrives at the end of the iterator, thus, there is no match.

Your untested code might be correct if you add that comparison with `== current_diff.end()`.
```c++
m_include.erase(
  std::remove_if(m_include.begin(), m_include.end(), [&](Diff& v) {
    // Remove values if it can't find matching symbols (i.e. iterator == std::end)
    return std::find_if(current_diff.begin(), current_diff.end(), [&](Diff& p) {
      return p.symbol == v.symbol;
    }) == current_diff.end();
  }
), m_include.end());
```

For clarity, we can keep `std::any_of`/`std::none_of` and it should give:
```c++
m_include.erase(
  std::remove_if(m_include.begin(), m_include.end(), [&](Diff& v) {
    // Remove values if it can't find matching symbols
    return std::none_of(current_diff.begin(), current_diff.end(), [&](Diff& p) {
      return p.symbol == v.symbol;
    });
  }
), m_include.end());
```

My initial point was to use `<algorithm>` when possible and if the logic allows it.

@TryTwo 
Was there a performance impact without a tmp swap?

Reply:
I definitely prefer the
```c++
m_include.erase(
  std::remove_if(m_include.begin(), m_include.end(), [&](Diff& v) {
    // Remove values if it can't find matching symbols
    return std::none_of(current_diff.begin(), current_diff.end(), [&](Diff& p) {
      return p.symbol == v.symbol;
    });
  }
), m_include.end());```

Tested. Appears to work fine. Using this solution.